### PR TITLE
Fix thread safety for collect in MeterProvider and AggregatorStores

### PR DESCRIPTION
**Pull request by cijothomas**
Adds lock to prevent
1. New instrument (i.e new aggregatorstore as we have aggregatorstore per instrument) from being added while Collect() is going on.
2. New timeseries within aggregatorstore from being added, while Collect() is going on.

Doing the easy fix, to unblock alpha1 release, as optimizing these would require more rearchitecting.

🔒 🔒

### CODE REVIEW

File: src/OpenTelemetry/Metrics/MeterProviderSdk.cs [hunk 1]
```diff
@@ -143,11 +149,11 @@ private MetricItem Collect(bool isDelta)
                 {
                     // Record all observable instruments
                     this.listener.RecordObservableInstruments();
+                    var dt = DateTimeOffset.UtcNow;
```
Comment (cijothomas) [inline src/OpenTelemetry/Metrics/MeterProviderSdk.cs:152]: unrelated change. This reads time once for entire Collect() for all the aggregators, instead of each aggregator reading time itself.

File: src/OpenTelemetry/CHANGELOG.md [hunk 2]
```diff
@@ -2,7 +2,9 @@
 
 ## Unreleased
 
-* Add Metrics support.
+* Add basic Metrics support with a single pipeline,
```
Comment (reyang) [inline src/OpenTelemetry/CHANGELOG.md:5]: The line wrapping is weird, is it aligned on column 80?

File: src/OpenTelemetry/Metrics/MeterProviderSdk.cs (hunk 1 above)
Comment (reyang) [reply src/OpenTelemetry/Metrics/MeterProviderSdk.cs:152]: This might be a problem for synchronous instruments. Not a blocker at this moment.

**Review by reyang — APPROVED** [review]

**Review by alanwest — APPROVED** [review]

File: src/OpenTelemetry/Metrics/AggregatorStore.cs [hunk 3]
```diff
@@ -154,22 +153,25 @@ internal void Update<T>(T value, ReadOnlySpan<KeyValuePair<string, object>> tags
...
                 {
-                    foreach (var metric in values.Value)
+                    foreach (var values in keys.Value)
                     {
-                        var m = metric.Collect(dt, isDelta);
-                        if (m != null)
+                        foreach (var metric in values.Value)
                         {
-                            collectedMetrics.Add(m);
+                            var m = metric.Collect(dt, isDelta);
```
Comment (victlu) [inline src/OpenTelemetry/Metrics/AggregatorStore.cs:170]: You may want to refactor to keep the metric.Collect() outside of the lock().

Not a blocker at this time.

**Review by victlu — APPROVED** [review]
